server/server_test.go

```diff
@@ -29,38 +32,476 @@ func performRequest(r http.Handler, method, path string, body io.Reader) (*httpt
+		expErr:  ``,
+	}, {
+		msg:     `nonexistent game`,
+		gameID:  `123`,
+		expCode: http.StatusNotFound,
+		expErr:  `Game not found`,
+	}}
+	cs, router := newServerAndRouter()
+	pIDs := seedPlayers(t, cs.db, 2)
+	for _, tc := range testCases {
+		// seed the db with a game
+		g, err := createGame(context.Background(), cs.db, pIDs)
+		require.NoError(t, err)
+		// make the request
+		var url string
+		if tc.gameID != `` {
+			url = `/game/` + tc.gameID
+		} else {
+			url = `/game/` + fmt.Sprintf(`%d`, g.ID)
+		}
+		w, err := performRequest(router, `GET`, url, nil)
+		require.NoError(t, err)
+		// verify
+		require.Equal(t, tc.expCode, w.Code)
+		if tc.expCode != http.StatusOK {
+			errMsg := readError(t, w)
+			assert.Equal(t, tc.expErr, errMsg)
+			continue
+		}
+		var game model.Game
+		readBody(t, w.Body, &game)
+		assert.Equal(t, g.ID, game.ID)
+	}
+}
+func TestGinGetPlayer(t *testing.T) {
+	testCases := []struct {
+		msg      string
+		playerID string
+		expCode  int
+		expErr   string
+	}{{
+		msg:      `good request`,
+		playerID: `p1`,
+		expCode:  http.StatusOK,
+		expErr:   ``,
+	}, {
+		msg:      `nonexistent player`,
+		playerID: `p9`,
+		expCode:  http.StatusNotFound,
+		expErr:   `Player not found`,
+	}}
+	cs, router := newServerAndRouter()
+	seedPlayers(t, cs.db, 2)
+	for _, tc := range testCases {
+		// make the request
+		url := `/player/` + tc.playerID
+		w, err := performRequest(router, `GET`, url, nil)
+		require.NoError(t, err)
+		// verify
+		require.Equal(t, tc.expCode, w.Code)
+		if tc.expCode != http.StatusOK {
+			errMsg := readError(t, w)
+			assert.Equal(t, tc.expErr, errMsg)
+			continue
+		}
+		var player model.Player
+		readBody(t, w.Body, &player)
+		assert.Equal(t, model.PlayerID(tc.playerID), player.ID)
+	}
+}
+
+func TestGinPostAction(t *testing.T) {
+	type request struct {
+		action  model.PlayerAction
+		expCode int
+		expErr  string
+	}
+	testCases := []struct {
+		msg  string
+		reqs []request
+	}{{
+		msg: `invalid action type`,
+		reqs: []request{{
+			action: model.PlayerAction{
+				ID:        `p1`,
+				Overcomes: 123,
+				Action:    ``,
+			},
+			expCode: http.StatusBadRequest,
+			expErr:  `Error: unknown action type`,
+		}},
+	}, {
+		msg: `try to do an action at an inappropriate time`,
+		reqs: []request{{
+			action: model.PlayerAction{
+				ID:        `p1`,
+				Overcomes: model.CountCrib,
+				Action: model.CountCribAction{
+					Pts: 2,
+				},
+			},
+			expCode: http.StatusBadRequest,
+			expErr:  `Error: Should overcome 0, but overcomes 5`,
+		}},
+	}, {
+		msg: `play a few actions`,
+		reqs: []request{{
+			action: model.PlayerAction{
+				ID:        `p1`,
+				Overcomes: model.DealCards,
+				Action: model.DealAction{
+					NumShuffles: 1,
+				},
+			},
+			expCode: http.StatusOK,
+			expErr:  ``,
+		}, {
+			// note: these BuildCribActions are attempting to put cards into the crib that
+			// these players likely don't have, but we're just testing the router here
```

Comment:
can we guarantee that these players don't have these cards?

Reply:
They _could_ have these cards because it's random - is it advantageous to us either way to prove that they do or don't have the cards? We really just care about testing the router

Reply:
yes -- we should care. Our unit tests should pass every time they're run if there are no code changes and no bugs. Allowing intermittent failures just because the player is randomly assigned cards is not ok:#

Reply:
This is testing that posting a game action to the server will return the specified HTTP code. Posting a `BuildCribAction` with cards that aren't in your hand triggers `notifyBlocking` which isn't an error, so still returns `http.StatusOK`; it's not possible to have intermittent failures for this test due to the cards in the hand.